### PR TITLE
feat(scim): Allow modification of roles for provisioned members

### PR DESCRIPTION
**Pull request by AniketDas-Tekky**
This change will allows users already provisioned through SCIM to have their roles modified. This requires the addition of `sentryOrgRole` in the response of any SCIM OrgMember APIs.

The tests may seem complicated, but the gist is
* If the user is currently managed by the IDP, then we want to do whatever the SCIM API requests
* If the user is not currently managed by the IDP, then we only want to modify the user if the IDP want to explicitly manage it (a non-blank sentryOrgRole).

### CODE REVIEW

File: src/sentry/scim/endpoints/members.py [hunk 1]
```diff
@@ -250,6 +251,83 @@ def delete(self, request: Request, organization, member) -> Response:
...
+                    and member.role.lower() == request.data["sentryOrgRole"].lower()
+                ):
+                    context = serialize(
+                        member, serializer=_scim_member_serializer_with_expansion(organization)
+                    )
+                    return Response(context, status=200)
+                requested_role = request.data["sentryOrgRole"].lower()
+                idp_role_restricted = True
+            else:
+                # Only reset role if current role isn't the default
```
Comment (cathyteng17) [inline src/sentry/scim/endpoints/members.py:304]: will we allow for the case where a member is currently idp:role-restricted and we want to edit them to have a blank role (and thereby remove the idp:role-restricted flag?)

Comment (AniketDas-Tekky) [reply src/sentry/scim/endpoints/members.py:304]: Yeah, my thinking is that if the user explicitly sets the role to blank, then they don't want this user's role to be fixed by the IDP

**Review by cathteng — APPROVED** [review]

**Review by cathteng — COMMENTED** [review]
i have been testing this with a custom integration in Okta (that i created through the wizard) and i'm unable to remove users (which uses PATCH). i haven't been able to find the root cause yet but i believe it's because of this from the Okta docs:

https://developer.okta.com/docs/reference/scim/scim-20/#update-the-user
> For any new OIN app integrations, all updates to a User object are handled using a PUT method request, except as noted in [Update a specific User (PATCH)](https://developer.okta.com/docs/reference/scim/scim-20/#update-a-specific-user-patch).

> For any custom app integrations created using the App Integration Wizard (AIW), all updates to a User object are handled using a PUT method request.

so i think when we actually update the Sentry Okta integration, SCIM should use PATCH for removing users and PUT for updating the member.

File: src/sentry/scim/endpoints/members.py [hunk 2]
```diff
@@ -251,6 +252,85 @@ def delete(self, request: Request, organization, member) -> Response:
...
+                if (
+                    member.role != organization.default_role
+                    and not member.flags["idp:role-restricted"]
```
Comment (nhsiehgit) [inline src/sentry/scim/endpoints/members.py:308]: If the role is restricted, shouldn't this api still modify the role? iiuc this api is what the IDP would use to provision roles right?

**Review by nhsiehgit — COMMENTED** [review]
Not quite sure if i understand the full context here 🤷 
mostly nit comments

File: src/sentry/scim/endpoints/members.py (hunk 2 above)
Comment (cathyteng17) [reply src/sentry/scim/endpoints/members.py:308]: this block is covering the case where the member isn't idp_role_restricted and you set the role in the IdP to blank --> nothing changes
https://github.com/getsentry/sentry/blob/d58431b5e553be7b6acd94be2e32fab1b52a7e93/tests/sentry/api/endpoints/test_scim_user_details.py#L148-L149

Comment (cathyteng17) [reply src/sentry/scim/endpoints/members.py:308]: also, here the member is not role-restricted. i'll add a comment to clarify though

**Pull request merged**